src/core/memory.d

```diff
@@ -1067,7 +1067,7 @@ void __delete(T)(ref T x) @system
     {
         static if (is(E == struct))
         {
-            foreach (ref e; x)
+            foreach_reverse (ref e; x)
```

Comment:
See https://github.com/dlang/druntime/blob/master/src/rt/lifetime.d#L1157 for currently happens with delete: finalizers are only called if it is a runtime allocated array (debatable), pointer offset is corrected for large arrays (GC.free is wrong otherwise), block cache is invalidated (not doing so might make alter operations unsafe).

Reply:
That seems to be an existing issue with `__delete`'s impementation, but out of scope of this PR.  Can you just file a bug report, so it can be dealt with properly?

Reply:
Probably we should just have `__delete` call `_d_delarray_t`.

Reply:
Can't do that as these C functions are supposed to go once the deprecation is over.

Reply:
> That seems to be an existing issue with `__delete`'s impementation, but out of scope of this PR. 

Fine with me.

> Can you just file a bug report, so it can be dealt with properly?

https://issues.dlang.org/show_bug.cgi?id=13558

> Probably we should just have `__delete` call `_d_delarray_t`.

That's what I thought, too.

Reply:
I think the implementation is still needed though; it just doesn't get called directly by the compiler.  We either need to move the implementation into `__delete` or have `__delete` call it.  I suppose we could also convert `_d_delarray_t` into a template to get rid of the dependency on `TypeInfo`.

Reply:
I think Andrei would want us to move the implementation into __delete as the general idea is to move into a D without any runtime.

Reply:
Ok, I'll look into that soon.